### DISCUSSION TITLE
Lunar moon change

### DISCUSSION THREAD

**Discussion by arumdyaharum** [cosinekitty/astronomy#309] [Q&A]
In astrology we have north node and south node that change the longitude almost every day. I tried code like this
```
function moonNode (date) {
  console.log(date);
  const node = Astronomy.SearchMoonNode(date);
  const eclip = Astronomy.EclipticGeoMoon(node.time);
  const type = node.kind = 1 ? "North node" : "South node"
  const degree = Math.floor(eclip.lon);
  const minutes = Math.floor((eclip.lon - degree) * 60) 
  console.log(type + ' ' + degree + ' ' + minutes)
  return eclip.lon;
}
```

the output was this
```
1998-04-03T00:00:00.000Z
North node 160 8
1998-01-05T14:13:56.903Z
North node 161 15
```

but after I check on the astrology empheris (https://www.astro.com/tmpd/c8yhfileyS5iO5-u1687025000/astro_d5eph_d.65025.59325.pdf), the actual degree (with adding 150 because it's in virgo) is this

```
1998-04-03
North Node 160 4
South Node 158 51

1998-01-05
North Node 162 16
South Node 163 31
```

How can I get the correct lunar node longitude like astrology empheris?

**Comment by cosinekitty** [comment]
The ephemeris you referenced from astro.com appears to be showing the Moon's ecliptic longitude at exactly 00:00 UTC. However, in the Javascript code you posted, you are searching for the next time the Moon crosses through the ecliptic plane, which is not the same time. In fact, it can be as many as 15 days later. Can you try adding this line of code?

```javascript
console.log(`Node at ${node.time}`);
```

This will show the time for which you are finding the Moon's ecliptic longitude.

I think what you meant to do is:

```javascript
const eclip = Astronomy.EclipticGeoMoon(date);
```

**Reply by arumdyaharum** [reply]
the console shows that the time is a little bit off

```
1998-04-03T00:00:00.000Z
AstroTime {
  date: 1998-04-08T09:49:18.274Z,
  ut: -633.0907607093271,
  tt: -633.0900306830393
}
North node 160 8

1998-01-05T14:13:56.903Z
AstroTime {
  date: 1998-01-16T15:52:45.923Z,
  ut: -714.8383573684636,
  tt: -714.8376288041575
}
North node 161 15
```

if I only use EclipticGeoMoon, I get current moon position, not current moon nodes position. I use ephemeris as an example to illustrate how lunar moon can change the degree everyday.

in astrology, the change of lunar nodes will determine when north node and south node are changing the sign in transit chart and which house it enters in natal chart. in human design, the accuracy of lunar nodes will determine outer variables which you can see from two bottom arrows on the chart. I'm trying to create human design chart which means I need the accuracy of the lunar nodes.

![image](https://github.com/cosinekitty/astronomy/assets/88584168/564d1903-0942-4c10-b879-0b355b88ce80)

![image](https://github.com/cosinekitty/astronomy/assets/88584168/2410ab47-e8d2-47b4-8538-0bce73afcc7a)

**Reply by cosinekitty** [reply]
Sorry, I'm not following what you are saying.

> the console shows that the time is a little bit off

The console log shows what I thought it would show, which is that you are searching for the next lunar node after a given time, which results in a later time. Like I said earlier, the next node crossing can occur anywhere from 0 to 15 days after the time you search from.

> if I only use EclipticGeoMoon, I get current moon position, not current moon nodes position.

You can pass any date/time to EclipticGeoMoon, not just the current time. When you say "current moon position", yes, you get the position of the Moon at whatever time you pass in as the parameter.

I can't figure out what you mean when you say "current moon nodes position". The node is a place where the moon crosses through the ecliptic plane, and that happens at a specific time. The node isn't always happening, so saying "current" confuses me. I suspect you are talking about a different concept using the word "node", but I can't figure out what that might be.

**Reply by arumdyaharum** [reply]
this is what I mean. I changed the chart in 1 day increment. you can see north node and south node is changing everyday.

![20230625_075900.gif](https://github.com/cosinekitty/astronomy/assets/88584168/ecbd83d8-1fc1-4f6f-94ea-317495a8cd5d)

**Reply by cosinekitty** [reply]
Can you define "north node" and "south node" as you are using it?

**Reply by arumdyaharum** [reply]
okay, I'm also not a native english speaker so I'll quote from articles I found.

> The points at which the orbital paths of the sun and the moon intersect are called lunar nodes. Because the moon completes an orbit around the earth in one month, in this period it intersects the sun’s orbital path twice, one time in a northerly direction called **the north or ascending lunar node**, and a half-month later in a southerly direction called **the south or descending node**. Within these monthly repetitions, the intersection point moves retrogressively through the ecliptic. - https://www.waldorflibrary.org/images/stories/articles/WJP2_donato.pdf

> Because the [orbital plane](https://en.wikipedia.org/wiki/Orbital_plane) of the Moon [precesses](https://en.wikipedia.org/wiki/Lunar_precession) in space, the lunar nodes [also precess](https://en.wikipedia.org/wiki/Nodal_precession) around the ecliptic - https://en.wikipedia.org/wiki/Lunar_node

basically north node is the northern ecliptic and south node is the southern ecliptic. because the moon orbit is changed so the angle of lunar nodes is changing too.

![image](https://github.com/cosinekitty/astronomy/assets/88584168/87621888-39bb-406e-8e83-bd2401839ca6)

![image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/96/Animation_of_Moon_orbit_around_Earth_-_Equatorial_view.gif/327px-Animation_of_Moon_orbit_around_Earth_-_Equatorial_view.gif)

**Reply by cosinekitty** [reply]
OK, this is good, because now I know we are talking about the same thing. The "north node" is the same thing I call an ascending node in Astronomy Engine, and the "south node" is the same thing I call the descending node. However, the first remark you made in this discussion still confuses me:

> In astrology we have north node and south node that change the longitude almost every day.

I don't know what you mean when you say the nodes change longitude almost every day. There are only 2 nodes, ascending and descending, and each happens only once during a given lunar orbit. So on average, there is one ascending node every 29 days and one descending node every 29 days. Neither of these can happen every day. Each occurs at a very specific time and longitude. Both happen at a fixed point at a fixed time, not changing every day. However, like the above quoted reference mentions, the longitude is different for each lunar month.

So I still feel like there is a problem where I don't understand what you are trying to do, or how Astronomy Engine's functionality is lacking for that case.

**Reply by arumdyaharum** [reply]
I tried to use two eclipse date around the date I enter and then I create percentage from those two date so I can get the longitude of the date I enter. I get pretty close degree to what I expected. here's my code 
```js
function lunarNodes (date) {
  const node = Astronomy.SearchMoonNode(date);
  let c = new Date(date);
  let flag = true;
  let nodeBefore = node;
  while(flag) {
    c.setDate(c.getDate() - 5);
    nodeBefore = Astronomy.SearchMoonNode(c);
    if (nodeBefore.time.date.getDate() !== node.time.date.getDate()) flag = false;
  }
  const eclip = Astronomy.EclipticGeoMoon(node.time);
  const eclipBefore = Astronomy.EclipticGeoMoon(nodeBefore.time);
  const q = Math.abs(date - nodeBefore.time.date);
  const d = Math.abs(node.time.date - nodeBefore.time.date);
  const fraction = (q/d);
  const currentElon = (eclip.lon - eclipBefore.lon) * (fraction / 100);
  return nodeBefore.kind > node.kind ? eclip.lon + 180 - currentElon : eclip.lon - currentElon;
}
```

**Reply by JoyL8896** [reply]
interesting

**Reply by venu-mallik** [reply]
@arumdyaharum I am not sure if the accuracy is enough for you, formulae is in page 65.
Formula : https://literature.hpcalc.org/community/astronomical-formulae.pdf

**Reply by arumdyaharum** [reply]
@vrworkers thank you for the reference. I don't have any astronomical background so it's quite confusing. though I find it helpful to learn about astronomy.

**Comment by arumdyaharum** [comment]
Hi again. I wanna update this discussion. I shared my human design program https://github.com/arumdyaharum/hdkit if you wanna know how I use this. if you run the program, you'll get gate, line, color, tone, and base. basically what it does is dividing gates into lines, lines into colors, etc. so the accuracy of the degree or longitude is crucial.

tbh I'm confused when I wrote [this answer of mine](https://github.com/cosinekitty/astronomy/discussions/309#discussioncomment-6310060). so I refactor it like this.
```js
function lunarNode (date) {
  const node = Astronomy.SearchMoonNode(date);
  const eclip = Astronomy.EclipticGeoMoon(node.time);
  const eclipLon = node.kind < 0 ? (eclip.lon + 180) % 360 : eclip.lon
  if (node.time.date === date) return eclipLon;
  let tempDate = new Date(date);
  let flag = true;
  let nodeBefore = node;
  while(flag) {
    tempDate.setDate(tempDate.getDate() - 10);
    nodeBefore = Astronomy.SearchMoonNode(tempDate);
    if (nodeBefore.time.date.getDate() !== node.time.date.getDate()) flag = false;
  }
  const eclipBefore = Astronomy.EclipticGeoMoon(nodeBefore.time);
  const eclipBeforeLon = nodeBefore.kind < 0 ? (eclipBefore.lon + 180) % 360 : eclipBefore.lon
  const d = Math.abs(node.time.date - nodeBefore.time.date);
  const a = Math.abs(node.time.date - date);
  const fraction = (a/d);
  let angle = eclipBeforeLon - eclipLon;
  angle = angle * fraction;
  const currentEclip = eclipLon + angle;
  // version 2 : with the constant of lunar nodes regression's rate 
  // const time = a / (365.25 * 24 * 60 * 60 * 1000);
  // const angle = 19.32 * time;
  // const currentEclip = eclipLon + angle;
  return currentEclip;
}
```
it's getting the lunar node's gates and line right, but if I wanna get the color and tone, the accuracy is not consistent especially before 2000. I know it's accurate or not by comparing my data to www.humdes.com data. lunar node's color and tone is used for creating the direction of bottom arrow. it also applies to the top arrow with sun's color and tone which if you see it's already accurate.

![image](https://github.com/cosinekitty/astronomy/assets/88584168/aa6d9406-d4b8-4d21-ae0a-ab511c59b004)
![image](https://github.com/cosinekitty/astronomy/assets/88584168/627b19df-6990-4d14-a4e2-fea9729751a2)
